memgpt/local_llm/json_parser.py

```diff
@@ -46,18 +46,114 @@ def add_missing_heartbeat(llm_json):
+
+    # Storage for the new JSON
+    new_json_parts = []
+
+    # Iterating through each character
+    for char in json_string:
+        if char == '"' and not escape:
+            in_string = not in_string
+            if not in_message:
+                # If we encounter a quote and are not in message, append normally
+                new_json_parts.append(char)
+        elif char == "\\" and not escape:
+            escape = True
+            new_json_parts.append(char)
+        else:
+            if escape:
+                escape = False
+            if in_message:
+                if char == "}":
+                    # Append the consolidated message and the closing characters then reset the flag
+                    new_json_parts.append('"{}"'.format("".join(message_content).replace("\n", " ")))
+                    new_json_parts.append(char)
+                    in_message = False
+                elif in_string or char.isalnum() or char.isspace() or char in ".',;:!":
+                    # Collect the message content, excluding structural characters
+                    message_content.append(char)
+            else:
+                # If we're not in message mode, append character to the output as is
+                new_json_parts.append(char)
+                if '"message":' in "".join(new_json_parts[-10:]):
+                    # If we detect "message": pattern, switch to message mode
+                    in_message = True
+                    message_content = []
+
+    # Joining everything to form the new JSON
+    repaired_json = "".join(new_json_parts)
+    return repaired_json
+
+
 def clean_json(raw_llm_output, messages=None, functions=None):
-    """Try a bunch of hacks to parse the data coming out of the LLM"""
+    from memgpt.utils import printd
 
+    """Try a bunch of hacks to parse the data coming out of the LLM"""
     try:
+        # printd("clean json runs:", raw_llm_output)
         data = json.loads(raw_llm_output)
     except json.JSONDecodeError:
         try:
+            printd("trying adding }")
             data = json.loads(raw_llm_output + "}")
         except json.JSONDecodeError:
             try:
-                data = extract_first_json(raw_llm_output + "}")
-            except:
-                raise
-
+                printd("trying adding }}")
+                data = json.loads(raw_llm_output + "}}")
+            except json.JSONDecodeError:
+                try:
+                    printd('trying adding "}}')
+                    data = json.loads(raw_llm_output + '"}}')
+                except json.JSONDecodeError:
+                    try:
+                        repaired = repair_json_string(raw_llm_output)
+                        printd("trying my repair json:", repaired)
+                        data = json.loads(repaired)
+                    except json.JSONDecodeError:
+                        try:
+                            repaired = repair_json_string(raw_llm_output)
```

Comment:
Should this one be `repair_even_worse_json`?

Reply:
New functionality coming to the API. https://platform.openai.com/docs/guides/text-generation/json-mode